### PR TITLE
Implementation of RPM sensor

### PR DESCRIPTION
**Pull request by roman-dvorak**
Hi, 
we have developed an RPM sensor for the measuring of the speed of rotary things (propeller, rotor, engine RPM, ...). Our colleague @slimonslimon assembled software for collecting data from the sensor and logging them in PX4 autopilot. 

 [TFRPM01](https://github.com/ThunderFly-aerospace/TFRPM01) sensor is based on [PCF8583](https://www.nxp.com/docs/en/data-sheet/PCF8583.pdf) counter. The sensor is fully open-source and manufacturing data are available on [GitHub](https://github.com/ThunderFly-aerospace/TFRPM01). The sensor board is equipped with a (two troughtpass) px4 standard I2C connectors, with led for quick diagnostic and with sensor 3 pin connector.
 
063A9592c

TFRPM01 does not contain the sensor itself, because it supports to connect many of types of sensors. For example a hall probe with magnets, some optical gate or reflecting light sensor. 

RPM is quite important to know for rotor based aircraft (helicopter, autogyro, ..). It can be also used for in-flight diagnostic of ESC (and engines). 

Snímek z 2020-01-24 01-23-34

So, with this pull-request, I would like to merge our driver with the master.

### CODE REVIEW

File: boards/px4/fmu-v5/default.cmake [hunk 1]
```diff
@@ -18,6 +18,7 @@ px4_add_board(
 	DRIVERS
 		adc
 		barometer # all available barometer drivers
+        rotor_frequency
```
Comment (dagar) [inline boards/px4/fmu-v5/default.cmake]:
```suggestion
		rotor_frequency
```

Can you keep the list sorted alphabetically?

File: src/drivers/rotor_frequency/pcf8583/pcf8583.cpp [hunk 2]
```diff
@@ -0,0 +1,702 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2013-2016 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/rotor_frequency/pcf8583/pcf8583.cpp]: Year

File: src/drivers/rotor_frequency/pcf8583/pcf8583.cpp [hunk 3]
```diff
@@ -0,0 +1,702 @@
...
+#include <uORB/uORB.h>
+#include <uORB/topics/rotor_frequency.h>
+
+#include <drivers/drv_hrt.h>
+#include <drivers/drv_sensor.h>
+
+
+/* Configuration Constants */
+#define PCF8583_BUS_DEFAULT 		      PX4_I2C_BUS_EXPANSION
+#define PCF8583_DEVICE_PATH	         "/dev/pcf8583"
```
Comment (dagar) [inline src/drivers/rotor_frequency/pcf8583/pcf8583.cpp]: The dev and ioctls can be gutted. Everything is done over uORB and parameters.

File: src/drivers/rotor_frequency/pcf8583/pcf8583.cpp [hunk 4]
```diff
@@ -0,0 +1,702 @@
...
+private:
+	int            _pool_interval; //Interval of reading counter and publishing frequency
+   int            _pool_interval_default; //Interval of reading counter and publishing frequency
+   float          _indicated_frequency;
+   float          _estimated_accurancy;
+   int            _count;
+   int            _reset_count;
+   int            _magnet_count;
+   uint64_t       _lastmeasurement_time;
+	work_s			_work{};
```
Comment (dagar) [inline src/drivers/rotor_frequency/pcf8583/pcf8583.cpp]: Please use the work queue within PX4. Here's an example - https://github.com/PX4/Firmware/blob/47cab3dba52520c4c3e9b42529152baf1148f1ee/src/drivers/lights/rgbled/rgbled.cpp#L63

File: boards/px4/fmu-v5/default.cmake (hunk 1 above)
Comment (roman-dvorak) [reply boards/px4/fmu-v5/default.cmake]: Sure, repaired in new commit

File: src/drivers/rotor_frequency/pcf8583/pcf8583.cpp (hunk 2 above)
Comment (roman-dvorak) [reply src/drivers/rotor_frequency/pcf8583/pcf8583.cpp]: done

**Comment by jinchengde** [comment]
that's very useful with hall sensor to monitor the status of gas engine

File: src/drivers/rotor_frequency/pcf8583/pcf8583.cpp (hunk 3 above)
Comment (slimonslimon) [reply src/drivers/rotor_frequency/pcf8583/pcf8583.cpp]: Done in last commit. But these defines remains for I2C class..

File: src/drivers/rotor_frequency/pcf8583/pcf8583.cpp (hunk 4 above)
Comment (slimonslimon) [reply src/drivers/rotor_frequency/pcf8583/pcf8583.cpp]: Done.

**Comment by roman-dvorak** [comment]
> that's very useful with hall sensor to monitor the status of gas engine, that will be better if support any hall sensor connect aux pin directly.

@jinchengde This counter can measure frequencies to 20kHz without any additional load to autopilot CPU. Therefore we decided to design this sensor. 

As I mentioned above, this PCB doesn't contain the hall sensor itself. And it supports the connection of external sensor. For example hall sensor, optical gate or other puls signal. And yes, it can be used for measuring engine RPM.

**Comment by slimonslimon** [comment]
I have renamed rotor_frequency to rpm in last commit. I think this name is better, because the driver is not limited only to rotors.

File: boards/px4/fmu-v5/default.cmake [hunk 5]
```diff
@@ -18,7 +18,7 @@ px4_add_board(
 	DRIVERS
 		adc
 		barometer # all available barometer drivers
-		batt_smbus
+        batt_smbus
```
Comment (dagar) [inline boards/px4/fmu-v5/default.cmake]:
```suggestion
		batt_smbus
```

File: msg/CMakeLists.txt [hunk 6]
```diff
@@ -160,6 +160,7 @@ set(msg_files
 	vtol_vehicle_status.msg
 	wheel_encoders.msg
 	wind_estimate.msg
+	rpm.msg
```
Comment (dagar) [inline msg/CMakeLists.txt]: We try to keep the list sorted alphabetically.

File: src/drivers/rpm/pcf8583/pcf8583.cpp [hunk 7]
```diff
@@ -0,0 +1,435 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2013-2020 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/rpm/pcf8583/pcf8583.cpp]:
```suggestion
 *   Copyright (c) 2020 PX4 Development Team. All rights reserved.
```

File: msg/CMakeLists.txt (hunk 6 above)
Comment (slimonslimon) [reply msg/CMakeLists.txt]: sry - fixed

File: src/drivers/rpm/pcf8583/pcf8583.cpp (hunk 7 above)
Comment (slimonslimon) [reply src/drivers/rpm/pcf8583/pcf8583.cpp]: okey

File: boards/px4/fmu-v5/default.cmake (hunk 5 above)
Comment (slimonslimon) [reply boards/px4/fmu-v5/default.cmake]: probably fixed.

**Pull request merged**